### ISSUE TITLE
Feature request: support for typeclasses in Ltac2

### ISSUE THREAD

**Issue by patrick-nicodemus** [coq/coq#18522]
Ltac2 currently does not appear to know about the Coq typeclass system.
```
From Ltac2 Require Import Ltac2.
Class myclass (A : Type) := {
    a : A
  }.

Instance nat_inst : myclass(nat) := {| a := 0%nat |}.

Theorem claim : myclass nat.
  Fail exact _.

The command has indeed failed with message: Cannot infer this placeholder of type "myclass nat" (no type class instance found).

   ltac1:(exact _).
Qed.

Definition def : myclass nat := ltac2:(refine '(_)).

Error:
The following term contains unresolved implicit arguments: 
  ?Goal
More precisely: 
- ?Goal: Cannot infer an existential variable of type "myclass nat" (no type class instance found).
```

I suggest that adding typeclasses directly to Ltac2 will allow developers to integrate Ltac2 more naturally into writing definitions for terms, especially by facilitating notation that suppresses implicit arguments.

**Comment by SkySkimmer** [comment]
It's not about typeclasses, it's missing bidirectionality. Fixed in 8.19 for exact.

**Comment by SkySkimmer** [comment]
Maybe we should have a short quotation for preterm like we have `'` for open_constr and a definition for preterm refine. eg
~~~coq
Ltac2 urefine (c:preterm) := Control.refine (fun () => pretype refine_flags c (Control.goal ())).

Goal ...
urefine preterm:(whatever). (* preterm:() is relatively heavy so it would be nice to have something shorter *)
~~~
IDK if it should be called urefine or if we should replace Notations.refine (with Control.refine staying as the low level primitive).

**Comment by patrick-nicodemus** [comment]
I have a vague understanding of what a preterm is but what are the most important properties of preterms that end users need to know about? Is there any discussion on zulip or GitHub where a coq developer explains what tactic users need to know about preterms?

**Comment by SkySkimmer** [comment]
(There's also some redundancy between `refine` and `unshelve eexact`.)

**Comment by patrick-nicodemus** [comment]
Can you explain how `Control.refine t` differs from `let x := t () in Control.refine (fun () => x)`?
The latter is just going to unify the type of the term with the goal and return all unsolved metavariables as new holes; as far as I can see, there should be nothing related to elaboration which occurs during the call to Control.refine, it's a full fledged term with all its remaining evars being tracked by the evar map of the proof state.

Suppose that `t` is doing something such as elaborating a preterm to an `open_constr`.
If both of these are running in the same proof context/proofview state,is there any additional context /changes the `refine` tactic supplies which could affect what `t` returns?

**Comment by patrick-nicodemus** [comment]
> Maybe we should have a short quotation for preterm like we have `'` for open_constr and a definition for preterm refine. eg
> 
> Ltac2 urefine (c:preterm) := Control.refine (fun () => pretype refine_flags c (Control.goal ())).
> 
> Goal ...
> urefine preterm:(whatever). (* preterm:() is relatively heavy so it would be nice to have something shorter *)
> 
> IDK if it should be called urefine or if we should replace Notations.refine (with Control.refine staying as the low level primitive).

I agree with this. Also, maybe, short notation for "open_constr_with_tc" to elaborate a term before it's passed to another tactic
```
Ltac2 Notation "otc" t(preterm) :=
  Constr.Pretype.pretype
    Constr.Pretype.Flags.open_constr_flags_with_tc
    Constr.Pretype.expected_without_type_constraint
    t.
```

**Comment by patrick-nicodemus** [comment]
I think the error messages displayed in the original post should be improved. Today I tried to use `ltac2:(Control.refine (fun () => 't)` in a Gallina term (i.e. not in proof mode) and ran into this problem again. The error message is identical to the error message you get in any context where typeclass search fails to resolve an instance, which may have a number of causes - the typeclass hint fails to unify for the goal, possibly for subtle reasons involving universe constraints, etc.
It would be better if the error message indicated that the typeclass search is happening in an empty typeclass context, or that no real typeclass search is happening at all.
